openjdk/src/main/java/org/conscrypt/NativeLibraryLoader.java

```diff
@@ -193,20 +184,57 @@ private static void load(String name, ClassLoader loader) {
         }
 
         if (url == null) {
-            // Fall back to normal loading of JNI stuff
-            loadLibrary(loader, name, false);
-            return;
+            return false;
         }
 
         int index = libname.lastIndexOf('.');
         String prefix = libname.substring(0, index);
         String suffix = libname.substring(index, libname.length());
+        File tmpFile = null;
+        try {
+            // Create a temporary file.
+            tmpFile = Platform.createTempFile(prefix, suffix, WORKDIR);
+            if (tmpFile.isFile() && tmpFile.canRead()
+                    && !NoexecVolumeDetector.canExecuteExecutable(tmpFile)) {
+                throw new IOException(MessageFormat.format("{0} exists but cannot be executed even "
+                                + "when execute permissions set; check volume for "
+                                + "\"noexec\" flag; use -D{1}=[path] to set native "
+                                + "working directory separately.",
+                        tmpFile.getPath(), WORK_DIR_PROPERTY_NAME));
+            }
+
+            // Copy the library from classpath to tmpFile
+            copyLibrary(url, tmpFile);
+
+            return loadLibrary(loader, tmpFile.getPath(), true, results);
+        } catch (IOException e) {
+            // Convert to an UnsatisfiedLinkError.
+            Throwable error = new UnsatisfiedLinkError(
+                    MessageFormat.format("Failed creating temp file ({0})",
+                            tmpFile)).initCause(e);
+            results.add(LoadResult.newFailureResult(name, true, false, error));
+            return false;
+        } finally {
+            // After we load the library it is safe to delete the file.
+            // We delete the file immediately to free up resources as soon as possible,
+            // and if this fails fallback to deleting on JVM exit.
+            if (tmpFile != null && (!DELETE_NATIVE_LIB_AFTER_LOADING || !tmpFile.delete())) {
```

Comment:
Having conditional side effects in an if statement like this seems like a recipe for future misunderstanding, so I'd prefer to do something more long-winded like

```java
if (tmpFile != null) {
  boolean deleted = false;
  if (DELETE_NATIVE_LIB_AFTER_LOADING) {
    deleted = tmpFile.delete();
  }
  if (!deleted) {
    tmpFile.deleteOnExit();
  }
}
```